File: src/day04/main.py
def check_card(line: str):
    _, numbers = line.split(':')
    winning_numbers, have_numbers = numbers.split('|')
    winning_numbers = set(int(x) for x in winning_numbers.split())
    have_numbers = set(int(x) for x in have_numbers.split())
    n = len(winning_numbers.intersection(have_numbers))
    v = (2 ** (n - 1)) if n else 0
    return (v, n)

def p2(cards: list[int]):
    hist = [1 for _ in cards]
    for i in range(len(hist)):
        n = hist[i]
        if n:
            begin = i + 1
            end = min(len(cards), i+cards[i]+1)
            for j in range(begin, end):
                hist[j] += n
    return sum(hist)
    
def main(input: str):
    cards = [check_card(line) for line in input.splitlines()]
    print(f'{sum(card[0] for card in cards)}') # 33950
    print(f'{p2([card[1] for card in cards])}') # 14814534

File: src/day04/main.rs
use std::collections::HashSet;

use shared::puzzle_input;

fn check_card(line: &str) -> (usize, usize) {
    let parse_set =
        |s: &str| -> HashSet<usize> { s.split(' ').filter_map(|s| s.parse().ok()).collect() };
    let (winning, have) = line.split_once(':').unwrap().1.split_once('|').unwrap();
    match parse_set(winning).intersection(&parse_set(have)).count() {
        0 => (0, 0),
        x => (2usize.pow((x - 1) as u32), x),
    }
}

fn p2(card_values: &[usize]) -> usize {
    let n = card_values.len();
    let mut multipliers = vec![1; n];
    for (i, card_value) in card_values.iter().enumerate() {
        if 0 < *card_value {
            let n_copies = multipliers[i];
            for x in multipliers.iter_mut().skip(i + 1).take(*card_value) {
                *x += n_copies;
            }
        }
    }
    multipliers.iter().sum()
}

fn main() {
    let input = puzzle_input!();
    let (values, n_matches): (Vec<_>, Vec<_>) = input.split('\n').map(check_card).unzip();
    println!("{}\n{}", values.iter().sum::<usize>(), p2(&n_matches)); // 33950 14814534
}
